File: 131.cpp
struct Solution {
    inline bool isPal(int i, int j, const string &s){
        while (i < j){
            if (s[i++] != s[j--]){
                return false;
            }
        }
        return true;
    }

    using seg = pair<int, int>;
    using part = vector<seg>;
    using collPart = vector<part>;

    vector<vector<string>> partition(const string &s) {
        int n = s.size();
        vector<int> occurIdx[26];
        for (int i = 0; i < n; ++i){
            occurIdx[s[i] - 'a'].push_back(i);
        }
        vector<int> prevOccur(n, -1);
        for (int i = 0; i < 26; ++i){
            for (int j = 1; j < occurIdx[i].size(); ++j){
                prevOccur[occurIdx[i][j]] = occurIdx[i][j - 1];
            }
        }
        vector<collPart> memo(n);
        memo[0].push_back({{make_pair(0, 1)}});
        for (int i = 1; i < n; ++i){
            for (int ptr = i; ptr != -1; ptr = prevOccur[ptr]){
                if (!isPal(ptr, i, s)){
                    continue;
                }
                if (ptr == 0){
                    memo[i].push_back({{make_pair(0, i + 1)}});
                } else {
                    for (part possibility: memo[ptr - 1]){
                        possibility.push_back({make_pair(ptr, i - ptr + 1)});
                        memo[i].push_back(possibility);
                    }
                }
            }
        }
        vector<vector<string>> res;
        for (part &p : memo[n - 1]){
            res.push_back({});
            for (seg &sg : p){
                res.back().push_back(s.substr(sg.first, sg.second));
            }
        }
        return res;
    }
};

File: 131.py
from functools import cache

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        occurIdx = [[] for _ in range(26)]
        for i, c in enumerate(s):
            occurIdx[ord(c) - ord('a')].append(i)
        prevOccur = [-1] * n
        for i in range(26):
            for j in range(1, len(occurIdx[i])):
                prevOccur[occurIdx[i][j]] = occurIdx[i][j - 1]
        memo = [[] for _ in range(n)]
        memo[0] = [[s[0]]]

        @cache
        def isPal(i, j):
            if i >= j:
                return True
            else:
                return s[i] == s[j] and isPal(i + 1, j - 1)

        for i in range(1, n):
            ptr = i
            while ptr != -1:
                if isPal(ptr, i):
                    if ptr == 0:
                        memo[i].append([s[:i + 1]])
                    else:
                        for part in memo[ptr - 1]:
                            memo[i].append(part + [s[ptr:i + 1]])
                ptr = prevOccur[ptr]
        return memo[n - 1]
